stor/dx.py

```diff
@@ -114,71 +157,672 @@ def temp_url(self, lifetime=300, filename=None):
+            )
+            if source.resource and to_rename:
+                moved_folder_path = folder_dest / source.name
+                dxpy.api.project_rename_folder(
+                    dest.canonical_project,
+                    input_params={
+                        'folder': moved_folder_path.resource,
+                        'name': dest.name
+                    }
+                )
+
+    def _movetree(self, dest):
+        """Moves the project or directory to a different folder within project.
+        Similar to _clonetree except it moves the source tree within the same project.
+
+        Like copytree, if the destination exists as a folder already, the source dir is
+        moved inside that dest folder with its original name.
+
+        The source cannot be the root directory.
+
+        Refer to _clonetree or copytree for detailed information.
+
+        Args:
+            dest (Path): The destination directory path within same project
+
+        Raises:
+            TargetExistsError: When destination directory already exists
+            DNAnexusError: When attempting to move across projects
+        """
+        if dest.canonical_project != self.canonical_project:
+            raise DNAnexusError('Cannot movetree across different projects')
+        if not self.resource:
+            raise DNAnexusError('Cannot move root folder within same project on DX')
+        if self == dest:
+            return
+        if dest == (self.drive+dest.project):  # need to convert dx://proj to dx://proj:
+            dest = dest + ':'
+
+        source = utils.remove_trailing_slash(self)
+        dest_is_dir = dest.isdir()
+        to_rename = True
+
+        if dest_is_dir or utils.has_trailing_slash(dest):
+            dest = dest / source.name
+            if dest.isdir():
+                raise TargetExistsError(
+                    'Destination path ({}) already exists, will not cause '
+                    'duplicate folders to exist. Remove the original first'
+                    .format(dest)
+                )
+            to_rename = False
+
+        folder_dest = dest.parent
+        if not dest_is_dir and folder_dest.resource:  # avoid calling makedirs_p on project
+            folder_dest.makedirs_p()
+
+        project_handler = dxpy.DXProject(source.canonical_project)
+
+        with _propagate_dx_exceptions():
+            project_handler.move_folder(
+                folder=source.resource,
+                destination=folder_dest.resource or '/'
+            )
+            if to_rename:
+                moved_folder_path = folder_dest / source.name
+                dxpy.api.project_rename_folder(
+                    dest.canonical_project,
+                    input_params={
+                        'folder': moved_folder_path.resource,
+                        'name': dest.name
+                    }
+                )
+        self.invalidate_cached_properties()
+
+    @_propagate_dx_exceptions()
```

Comment:
We wanted to use a with block to limit the scope where the exceptions could be raised by. Since this function has only one line, I figured it looked more elegant this way. I would like to keep it please :)

Reply:
okay fine